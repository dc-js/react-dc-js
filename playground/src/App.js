import React, { useEffect, useState } from 'react'
import crossfilter from 'crossfilter'
import * as d3 from 'd3'
import csv from './ndx.csv'
import { BarChart, PieChart, BubbleChart } from 'react-dc-js'

import 'dc/dist/style/dc.css'
import './App.css'

const numberFormat = d3.format('.2f')

function App() {
  const [cx, setCx] = useState(null)

  useEffect(() => {
    ;(async () => {
      const data = await d3.csv(csv)
      const dateFormatSpecifier = '%m/%d/%Y'
      const dateFormatParser = d3.timeParse(dateFormatSpecifier)
      data.forEach(d => {
        d.dd = dateFormatParser(d.date)
        d.month = d3.timeMonth(d.dd) // pre-calculate month for better performance
        d.close = +d.close // coerce to number
        d.open = +d.open
      })
      const cx = crossfilter(data)
      setCx(cx)
    })()
  }, [])

  if (!cx) {
    return <p>Loading Data...</p>
  }

  const moveMonths = cx.dimension(d => d.month)
  const volumeByMonthGroup = moveMonths
    .group()
    .reduceSum(d => d.volume / 500000)

  const gainOrLoss = cx.dimension(d => (d.open > d.close ? 'Loss' : 'Gain'))
  const gainOrLossGroup = gainOrLoss.group()

  const yearlyDimension = cx.dimension(d => d3.timeYear(d.dd).getFullYear())
  const yearlyPerformanceGroup = yearlyDimension.group().reduce(
    (p, v) => {
      ++p.count
      p.absGain += v.close - v.open
      p.fluctuation += Math.abs(v.close - v.open)
      p.sumIndex += (v.open + v.close) / 2
      p.avgIndex = p.sumIndex / p.count
      p.percentageGain = p.avgIndex ? (p.absGain / p.avgIndex) * 100 : 0
      p.fluctuationPercentage = p.avgIndex
        ? (p.fluctuation / p.avgIndex) * 100
        : 0
      return p
    },
    (p, v) => {
      --p.count
      p.absGain -= v.close - v.open
      p.fluctuation -= Math.abs(v.close - v.open)
      p.sumIndex -= (v.open + v.close) / 2
      p.avgIndex = p.count ? p.sumIndex / p.count : 0
      p.percentageGain = p.avgIndex ? (p.absGain / p.avgIndex) * 100 : 0
      p.fluctuationPercentage = p.avgIndex
        ? (p.fluctuation / p.avgIndex) * 100
        : 0
      return p
    },
    () => ({
      count: 0,
      absGain: 0,
      fluctuation: 0,
      fluctuationPercentage: 0,
      sumIndex: 0,
      avgIndex: 0,
      percentageGain: 0,
    })
  )

  return (
    <div className="App">
      <header className="App-header">
        <BubbleChart
          width={990}
          height={250}
          transitionDuration={1500}
          margins={{ top: 10, right: 50, bottom: 30, left: 40 }}
          dimension={yearlyDimension}
          group={yearlyPerformanceGroup}
          colors={d3.schemeRdYlGn[9]}
          colorDomain={[-500, 500]}
          colorAccessor={d => d.value.absGain}
          keyAccessor={p => p.value.absGain}
          valueAccessor={p => p.value.percentageGain}
          radiusValueAccessor={p => p.value.fluctuationPercentage}
          maxBubbleRelativeSize={0.3}
          x={d3.scaleLinear().domain([-2500, 2500])}
          y={d3.scaleLinear().domain([-100, 100])}
          r={d3.scaleLinear().domain([0, 4000])}
          elasticY={true}
          elasticX={true}
          yAxisPadding={100}
          xAxisPadding={500}
          renderHorizontalGridLines={true}
          renderVerticalGridLines={true}
          xAxisLabel={'Index Gain'}
          yAxisLabel={'Index Gain %'}
          renderLabel={true}
          label={p => p.key}
          renderTitle={true}
          title={p =>
            [
              p.key,
              `Index Gain: ${numberFormat(p.value.absGain)}`,
              `Index Gain in Percentage: ${numberFormat(
                p.value.percentageGain
              )}%`,
              `Fluctuation / Index Ratio: ${numberFormat(
                p.value.fluctuationPercentage
              )}%`,
            ].join('\n')
          }
        />
        <PieChart
          dimension={gainOrLoss}
          group={gainOrLossGroup}
          width={180}
          height={180}
          radius={80}
        />
        <BarChart
          dimension={moveMonths}
          group={volumeByMonthGroup}
          width={990}
          height={40}
          margins={{ top: 0, right: 50, bottom: 20, left: 40 }}
          centerBar={true}
          gap={1}
          x={d3
            .scaleTime()
            .domain([new Date(1985, 0, 1), new Date(2012, 11, 31)])}
          round={d3.timeMonth.round}
          alwaysUseRounding={true}
          xUnits={d3.timeMonths}
        />
      </header>
    </div>
  )
}

export default App
