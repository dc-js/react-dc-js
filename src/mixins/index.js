// @flow

import type { BaseMixinProps } from './base-mixin'
import type { StackMixinProps } from './stack-mixin'

export type MixinProps = BaseMixinProps & StackMixinProps

export { baseMixin } from './base-mixin'
export { stackMixin } from './stack-mixin'
export type { BaseMixinProps, StackMixinProps }
