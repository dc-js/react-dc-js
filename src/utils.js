// @flow

import type { MixinProps } from './mixins'

export function combineMixins(mixins: Array<(any, MixinProps) => any> = []) {
  return function (chart: any, props: MixinProps) {
    // Transforms the received props into corresponding chart methods
    // Ensures that we can map only a subset of props to mixin methods
    const unhandledProps: any = mixins.reduce(
      (value, mixin) => mixin(chart, value),
      props
    )
    // Invoke the remaining generic props to corresponding methods
    Object.keys(unhandledProps).forEach(
      prop => prop in chart && chart[prop](unhandledProps[prop])
    )
  }
}
