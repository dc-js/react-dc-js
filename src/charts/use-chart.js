import { useEffect, useRef } from 'react'
import { InvalidStateException } from 'dc'

import { baseMixin } from '../mixins'
import { combineMixins } from '../utils'

export function useChart(chartFunc, props, mixin = combineMixins([baseMixin])) {
  const chartRef = useRef(null)
  const chart = useRef(null)
  const isRendered = useRef(false)

  useEffect(() => {
    chart.current = chartFunc(chartRef.current)
    isRendered.current = false
  }, [chartFunc])

  useEffect(() => {
    if (!chart.current) return
    mixin && mixin(chart.current, props)
    if (!isRendered.current) {
      try {
        chart.current.render()
        isRendered.current = true
      } catch (e) {
        if (!(e instanceof InvalidStateException)) throw e
      }
    } else {
      chart.current.redraw()
    }
  }, [chartFunc, mixin, props])

  return [chart, chartRef]
}
