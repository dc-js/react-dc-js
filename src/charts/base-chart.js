// @flow

import React from 'react'
import { baseMixin } from '../mixins'
import type { MixinProps } from '../mixins'
import { combineMixins } from '../utils'
import { useChart } from './use-chart'

/**
 * A helper function for generic charts.
 * Allows the hook to remain modular and not return HTML.
 * @param props
 * @param chartFunc
 * @param mixins
 */
export function BaseChart(
  props: $Shape<MixinProps>,
  chartFunc: any => mixed,
  mixins: Array<(any, MixinProps) => mixed> = [baseMixin]
) {
  const [, chartRef] = useChart(chartFunc, props, combineMixins(mixins))

  return <div ref={chartRef} />
}
