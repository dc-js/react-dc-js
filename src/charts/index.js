import {
  barChart,
  boxPlot,
  bubbleChart,
  bubbleOverlay,
  cboxMenu,
  compositeChart,
  dataCount,
  dataTable,
  geoChoroplethChart,
  heatMap,
  htmlLegend,
  legend,
  numberDisplay,
  pieChart,
  rowChart,
  scatterPlot,
  selectMenu,
  seriesChart,
  sunburstChart,
  textFilterWidget,
} from 'dc'
import { baseMixin, stackMixin } from '../mixins'
import { BaseChart } from './base-chart'
import ChartContext from './chart-context'
import { LineChart } from './line-chart'
import { useChart } from './use-chart'

export const BarChart = BaseChart(barChart, [baseMixin, stackMixin])
export const BoxPlot = BaseChart(boxPlot)
export const BubbleChart = BaseChart(bubbleChart)
export const PieChart = BaseChart(pieChart)
export const BubbleOverlay = BaseChart(bubbleOverlay)
export const CboxMenu = BaseChart(cboxMenu)
export const CompositeChart = BaseChart(compositeChart)
export const DataCount = BaseChart(dataCount)
export const DataTable = BaseChart(dataTable)
export const GeoChoroplethChart = BaseChart(geoChoroplethChart)
export const Heatmap = BaseChart(heatMap)
export const HtmlLegend = BaseChart(htmlLegend)
export const Legend = BaseChart(legend)
export const NumberDisplay = BaseChart(numberDisplay)
export const RowChart = BaseChart(rowChart)
export const ScatterPlot = BaseChart(scatterPlot)
export const SelectMenu = BaseChart(selectMenu)
export const SeriesChart = BaseChart(seriesChart)
export const SunburstChart = BaseChart(sunburstChart)
export const TextFilterWidget = BaseChart(textFilterWidget)
export { useChart, LineChart, ChartContext }
