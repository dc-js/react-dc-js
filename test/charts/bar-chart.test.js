import React from 'react'
import { render } from '@testing-library/react'
import crossfilter from 'crossfilter'
import * as d3 from 'd3'
import * as dc from 'dc'
import { BarChart } from 'src/charts'
import testSet from '../countries.json'

describe('bar chart', () => {
  it('should match snapshot', () => {
    const ndx = crossfilter(testSet)
    const dimension = ndx.dimension(d => d.title)
    const { container } = render(
      <BarChart
        dimension={dimension}
        group={dimension.group().reduceCount()}
        x={d3.scaleBand()}
        xUnits={dc.units.ordinal}
        transitionDuration={0}
      />
    )
    expect(container).toMatchSnapshot()
  })
})
