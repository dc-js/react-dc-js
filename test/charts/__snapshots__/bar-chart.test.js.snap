// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bar chart should match snapshot 1`] = `
<div>
  <div
    class="dc-chart"
  >
    <svg
      height="200"
      width="200"
    >
      <g>
        <g
          class="chart-body"
          clip-path="url(http://localhost/#dc-chart1-clip)"
          transform="translate(30, 10)"
        >
          <g
            class="stack _0"
          >
            <rect
              class="bar"
              fill="#1f77b4"
              height="80"
              width="38"
              x="1"
              y="80"
            >
              <title>
                England: 1
              </title>
            </rect>
            <rect
              class="bar"
              fill="#1f77b4"
              height="160"
              width="38"
              x="41"
              y="0"
            >
              <title>
                Sweden: 2
              </title>
            </rect>
            <rect
              class="bar"
              fill="#1f77b4"
              height="80"
              width="38"
              x="81"
              y="80"
            >
              <title>
                USA: 1
              </title>
            </rect>
          </g>
        </g>
        <g
          class="axis x"
          fill="none"
          font-family="sans-serif"
          font-size="10"
          text-anchor="middle"
          transform="translate(30,170)"
        >
          <path
            class="domain"
            d="M0.5,6V0.5H120.5V6"
            stroke="currentColor"
          />
          <g
            class="tick"
            opacity="1"
            transform="translate(20,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              England
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(60,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              Sweden
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(100,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              USA
            </text>
          </g>
        </g>
        <g
          class="axis y"
          fill="none"
          font-family="sans-serif"
          font-size="10"
          text-anchor="end"
          transform="translate(30,10)"
        >
          <path
            class="domain"
            d="M-6,160.5H0.5V0.5H-6"
            stroke="currentColor"
          />
          <g
            class="tick"
            opacity="1"
            transform="translate(0,160.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              0.0
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,144.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              0.2
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,128.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              0.4
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,112.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              0.6
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,96.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              0.8
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,80.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              1.0
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,64.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              1.2
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,48.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              1.4
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,32.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              1.6
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,16.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              1.8
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,0.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              2.0
            </text>
          </g>
        </g>
      </g>
      <defs>
        <clippath
          id="dc-chart1-clip"
        >
          <rect
            height="160"
            transform="translate(-0, -0)"
            width="120"
          />
        </clippath>
      </defs>
    </svg>
  </div>
</div>
`;
